### PR TITLE
Improving swift-objc interop

### PR DESCRIPTION
**Pull request by mikelikespie**
(Depends on #2313 being merged)

This should allow one to import swift libraries using clang import syntax, etc.

 * Modulemap interop for Skylark. Before, objc would never look at
  generated module maps since it only included values from
  `TOP_LEVEL_MODULE_MAP` which could not be set in swift
    - Split TOP_LEVEL_MODULE_MAP into two properties, TOP_LEVEL_MODULE_MAP (now an Artifact)
    and TOP_LEVEL_MODULE_NAME. This lets these be retrieved and set
    within skylark since there was no CppModuleMap converter.
    - Made it so when these were set from skylark, they weren't
    propagated to transitive dependencies
 * Added test to verify one could import swift using `@import` syntax
 * Added skylark helper to create actions to generate modulemaps. This
   allows swift modules to be created with modulemaps by default which
   import the compatibility header

### CODE REVIEW

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by laszlocsomor** [comment]
Sorry again about the atrocious delay in looking at this PR... assigning reviewers now.

File: tools/build_defs/apple/swift.bzl [hunk 1]
```diff
@@ -340,7 +373,16 @@ def _swift_library_impl(ctx):
   output_module = ctx.new_file(objs_outputs_path + module_name + ".swiftmodule")
 
   # These filenames are guaranteed to be unique, no need to scope.
-  output_header = ctx.new_file(ctx.label.name + "-Swift.h")
+  output_header = ctx.new_file(module_name + "-Swift.h")
+
+  # Unextended modulemap is used when compiling ourselves. It does not contain the .Swift clang submodule.
+  unextended_modulemap = ctx.new_file(ctx.label.name + "-unextended.modulemap")
```
Comment (dmishe) [inline tools/build_defs/apple/swift.bzl]: Why is this needed, does it trip the compiler if there's a .Swift submodule?

File: tools/build_defs/apple/swift.bzl [hunk 2]
```diff
@@ -340,7 +373,16 @@ def _swift_library_impl(ctx):
...
+  modulemap_action(ctx, output=unextended_modulemap, **_modulemap_info(module_name, output_header, objc_providers, unextended=True))
+
+  # This is the modulemap our dependents use. It includes our -Swift header.
+  output_modulemap = ctx.new_file(ctx.label.name + ".modulemap")
```
Comment (dmishe) [inline tools/build_defs/apple/swift.bzl]: Is there a good reason to use modules for this, instead of a simple #import?

File: tools/build_defs/apple/swift.bzl [hunk 3]
```diff
@@ -382,8 +424,11 @@ def _swift_library_impl(ctx):
   ctx.file_action(output=swiftc_output_map_file, content=swiftc_output_map.to_json())
 
   args = _swift_xcrun_args(ctx) + ["swiftc"] + swiftc_args(ctx)
+  args += sorted(set(["-I" + d.dirname for d in dep_modules + [output_module]]))
+
   args += [
-      "-I" + output_module.dirname,
+      "-Xcc", "-fmodule-map-file=" + unextended_modulemap.path,
```
Comment (dmishe) [inline tools/build_defs/apple/swift.bzl]: If I understand this right, instead of loading a module map generated by each objc_ dependency, you generate a huge map, that includes all deps right in this target, correct?

File: tools/build_defs/apple/swift.bzl [hunk 4]
```diff
@@ -382,8 +424,11 @@ def _swift_library_impl(ctx):
...
+      "-import-underlying-module",
```
Comment (dmishe) [inline tools/build_defs/apple/swift.bzl]: This will error out if there's no module with the same name as this swift_library. In most of the cases there won't be one.

In addition, using this with a giant modulemap would mean that all Objective-C symbols in the deps of this module are exposed to the clients, right?

**Comment by iirina** [comment]
@mikelikespie can you followup on the review?

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by mikelikespie** [comment]
Revisiting this patch... will have updates in a few days

**Comment by mhlopko** [comment]
Hi @mikelikespie, do I read this right that we're waiting for your update? No need to ping reviewers?

**Comment by mhlopko** [comment]
Friendly ping @mikelikespie

**Comment by bazel-io** [comment]
Can one of the admins verify this patch?

**Comment by mikelikespie** [comment]
@mhlopko I can close this for now .Thanks.
